src/logged/mod.rs

```diff
@@ -0,0 +1,511 @@
+#![doc = include_str!("README.md")]
+
+use super::base64_serde;
+use crate::commit_log::{
+    CommitLog, ConsumerOffset, ConsumerRecord, Header, PartitionOffsets, ProducedOffset,
+    ProducerError, ProducerRecord, Subscription, Topic,
+};
+use async_stream::stream;
+use async_trait::async_trait;
+use bytes::{Buf, BytesMut};
+use chrono::{DateTime, Utc};
+use log::{trace, warn};
+use metrics::increment_counter;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    pin::Pin,
+    time::Duration,
+};
+use tokio::{
+    fs::{self, File},
+    io::{AsyncReadExt, AsyncWriteExt},
+    sync::{mpsc, oneshot},
+    time,
+};
+use tokio_stream::Stream;
+use tokio_util::codec::Decoder;
+
+const CONSUMER_QUEUE_SIZE: usize = 10;
+const PRODUCER_QUEUE_SIZE: usize = 10;
+const TOPIC_FILE_CONSUMER_POLL: Duration = Duration::from_secs(1);
+const TOPIC_LABEL: &str = "topic";
+
+/// A commit log implementation that uses the file system as its
+/// backing store.
+///
+/// Notes:
+/// 1. Partition values cannot be non-zero.
+/// 2. The number of subscriptions of a topic will translate to
+///    the number of tasks that are spawned, along with their
+///    associated resources.
+#[derive(Clone)]
+pub struct FileLog {
+    producer_tx: mpsc::Sender<(
+        ProducerRecord,
+        oneshot::Sender<Result<ProducedOffset, ProducerError>>,
+    )>,
+    root_path: PathBuf,
+}
+
+#[derive(Clone, Deserialize, Debug, Eq, PartialEq, Serialize)]
+struct StorableRecord {
+    version: u32,
+    headers: Option<Vec<Header>>,
+    timestamp: Option<DateTime<Utc>>,
+    key: u64,
+    #[serde(with = "base64_serde")]
+    value: Vec<u8>,
+    offset: u64,
+}
+
+impl FileLog {
+    pub fn new(root_path: PathBuf) -> Self {
+        let retained_root_path = root_path.clone();
+
+        let (producer_tx, mut producer_rx) = mpsc::channel::<(
+            ProducerRecord,
+            oneshot::Sender<Result<ProducedOffset, ProducerError>>,
+        )>(PRODUCER_QUEUE_SIZE);
+
+        tokio::spawn({
+            async move {
+                let mut topic_files: HashMap<Topic, File> = HashMap::new();
```

Comment:
Turbo fish instead of type annotation?

Reply:
I don’t think that’s possible. The new function does take generics… unless I’m missing something here.

Reply:
No, you are right, my fault. Is the type annotation needed at all?

Reply:
Yeah. Needed.